### ISSUE TITLE
CDF function and model p

### ISSUE THREAD

**Issue by june1819** [fab-jul/L3C-PyTorch#26]
1.  In picture of paper, the picture shows that encode z(s) with model p(z(s) | f(s+1)) . In encoding details, it says encode each channel c of z(s) with C(z(s) | f(s+1)). Why?
2.  What is the relationship between CDF function and model p？

**Comment by kuaidiyuan** [comment]
I also have doubts about this. How to ensure that the CDF is an incremental function in encode? In the process of running forward on pytorch1.7, I found that the CDF function obtained from some images is not incremental, leading to wrong coding.

**Comment by fab-jul** [comment]
@june1819 
1. We encode with the CDF, and we encode channel per channel. The CDF and PDF are calculatable from each other, and in the paper we use the simpler notation (with p). The fact that we use a CDF for arithmetic coding is because it allows fast lookups of intervals (needed for arithmetic coding). 
2. CDF is the integrated p.

@kuaidiyuan The CDF obtained by the model is not always incremental, correct. That's why we have [`_renorm_cast_cdf_(cdf, precision)`](https://github.com/fab-jul/L3C-PyTorch/blob/f0c7aaaad88af0df574e9dded263606337414def/src/torchac/torchac.py#L203). For some more details on this, see the [torchac README](https://github.com/fab-jul/torchac#normalized-vs-unnormalized--floating-point-vs-integer-cdfs). This should work on all images. Can you share the image that led to wrong encoding?

**Comment by kuaidiyuan** [comment]
@fab-jul 
[click](https://github.com/kuaidiyuan/cour.git)
I put the weight of image and training here.
